File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/EditUserRoleFragment.java
package com.example.gemainventory.ui.settings;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.gemainventory.R;
import com.example.gemainventory.api.RetrofitClient;
import com.example.gemainventory.model.Usuario;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.imageview.ShapeableImageView;

import java.util.HashMap;
import java.util.Map;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class EditUserRoleFragment extends Fragment {

    private String targetUserId;
    private Usuario targetUser;
    private int currentUserRol;

    private android.widget.TextView tvName, tvEmail;
    private AutoCompleteTextView spinnerRole;
    private MaterialButton btnSave;

    private final String[] roles = { "ADMINISTRADOR", "SUPERVISOR", "VENDEDOR", "REPARTIDOR", "ALMACENISTA" };
    private final Map<String, Integer> roleMap = new HashMap<>();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            targetUserId = getArguments().getString("userId");
        }

        SharedPreferences prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE);
        currentUserRol = prefs.getInt("user_rol", -1);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_edit_user_role, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        roleMap.put("ADMINISTRADOR", 2);
        roleMap.put("SUPERVISOR", 3);
        roleMap.put("VENDEDOR", 4);
        roleMap.put("REPARTIDOR", 5);
        roleMap.put("ALMACENISTA", 6);

        view.findViewById(R.id.btn_back_edit_role).setOnClickListener(v -> requireActivity().onBackPressed());

        tvName = view.findViewById(R.id.tv_edit_user_name);
        tvEmail = view.findViewById(R.id.tv_edit_user_email);
        spinnerRole = view.findViewById(R.id.spinner_edit_user_role);
        btnSave = view.findViewById(R.id.btn_update_user_role);

        ArrayAdapter<String> adapter = new ArrayAdapter<>(requireContext(), android.R.layout.simple_dropdown_item_1line,
                roles);
        spinnerRole.setAdapter(adapter);

        fetchTargetUser();

        btnSave.setOnClickListener(v -> updateRole());
    }

    private void fetchTargetUser() {
        RetrofitClient.INSTANCE.getInstance().getUsuarioById(targetUserId).enqueue(new Callback<Usuario>() {
            @Override
            public void onResponse(Call<Usuario> call, Response<Usuario> response) {
                if (response.isSuccessful() && response.body() != null) {
                    targetUser = response.body();
                    displayUserInfo();
                    
                    // Si el usuario objetivo es el Propietario, deshabilitar edición
                    if (targetUser.getIdRol() != null && targetUser.getIdRol() == 1) {
                        btnSave.setEnabled(false);
                        spinnerRole.setEnabled(false);
                        Toast.makeText(getContext(), "El rol de Propietario es inmutable", Toast.LENGTH_LONG).show();
                    }
                } else {
                    Toast.makeText(getContext(), "Error al obtener datos del usuario", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Usuario> call, Throwable t) {
                Toast.makeText(getContext(), "Error de red al cargar usuario", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void displayUserInfo() {
        if (targetUser != null) {
            tvName.setText(targetUser.getNombre());
            tvEmail.setText(targetUser.getCorreo());

            String currentRoleName = "Desconocido";
            if (targetUser.getIdRol() != null) {
                for (Map.Entry<String, Integer> entry : roleMap.entrySet()) {
                    if (entry.getValue().equals(targetUser.getIdRol())) {
                        currentRoleName = entry.getKey();
                        break;
                    }
                }
            }
            spinnerRole.setText(currentRoleName, false);
        }
    }

    private void updateRole() {
        String selectedRoleName = spinnerRole.getText().toString();
        Integer newRoleId = roleMap.get(selectedRoleName);

        if (targetUser == null || newRoleId == null)
            return;

        // Validación de UI (ya protegida en backend también)
        if (currentUserRol == 2 && targetUser.getIdRol() != 3) {
            Toast.makeText(getContext(), "No tienes rango para editar a este usuario", Toast.LENGTH_LONG).show();
            return;
        }

        targetUser.setIdRol(newRoleId);

        RetrofitClient.INSTANCE.getInstance().actualizarUsuario(targetUserId, targetUser)
                .enqueue(new Callback<Usuario>() {
                    @Override
                    public void onResponse(Call<Usuario> call, Response<Usuario> response) {
                        if (response.isSuccessful()) {
                            Toast.makeText(getContext(), "Rol actualizado correctamente", Toast.LENGTH_SHORT).show();
                            requireActivity().onBackPressed();
                        } else if (response.code() == 403) {
                            // Ahora el backend manda mensajes específicos en el cuerpo o podemos usar el general
                            Toast.makeText(getContext(), "Error de Jerarquía: No tienes permisos para este cambio", Toast.LENGTH_LONG).show();
                        } else {
                            Toast.makeText(getContext(), "Error del Servidor: " + response.code(), Toast.LENGTH_SHORT).show();
                        }
                    }

                    @Override
                    public void onFailure(Call<Usuario> call, Throwable t) {
                        Toast.makeText(getContext(), "Fallo de conexión", Toast.LENGTH_SHORT).show();
                    }
                });
    }
}

File: GEMA_andorid/GEMAINVENTORY/app/src/main/java/com/example/gemainventory/ui/settings/EditUserRoleFragment.kt
package com.example.gemainventory.ui.settings

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.runtime.*
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import com.example.gemainventory.api.RetrofitClient
import com.example.gemainventory.model.Usuario
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class EditUserRoleFragment : Fragment() {

    private var targetUserId: String? = null
    private var currentUserRol: Int = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        targetUserId = arguments?.getString("userId")
        
        val prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE)
        currentUserRol = prefs.getInt("user_rol", -1)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val prefs = requireContext().getSharedPreferences("GemaPrefs", Context.MODE_PRIVATE)
        val isDarkMode = prefs.getBoolean("DarkMode", false)

        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                var targetUser by remember { mutableStateOf<Usuario?>(null) }
                var isLoading by remember { mutableStateOf(true) }

                LaunchedEffect(Unit) {
                    fetchTargetUser { user ->
                        targetUser = user
                        isLoading = false
                    }
                }

                EditUserRoleScreen(
                    darkTheme = isDarkMode,
                    user = targetUser,
                    currentUserRole = currentUserRol,
                    isLoading = isLoading,
                    onBackClick = { requireActivity().onBackPressed() },
                    onUpdateClick = { newRoleId ->
                        updateRole(targetUser, newRoleId)
                    }
                )
            }
        }
    }

    private fun fetchTargetUser(onResult: (Usuario?) -> Unit) {
        val id = targetUserId ?: return
        RetrofitClient.instance.getUsuarioById(id).enqueue(object : Callback<Usuario> {
            override fun onResponse(call: Call<Usuario>, response: Response<Usuario>) {
                if (response.isSuccessful) {
                    onResult(response.body())
                } else {
                    Toast.makeText(context, "Error al obtener datos", Toast.LENGTH_SHORT).show()
                    onResult(null)
                }
            }

            override fun onFailure(call: Call<Usuario>, t: Throwable) {
                Toast.makeText(context, "Error de red", Toast.LENGTH_SHORT).show()
                onResult(null)
            }
        })
    }

    private fun updateRole(user: Usuario?, newRoleId: Int) {
        val id = targetUserId ?: return
        val currentUser = user ?: return

        // Validación de jerarquía (espejo de la lógica anterior)
        if (currentUserRol == 2 && currentUser.idRol != 3) {
            Toast.makeText(context, "No tienes rango para editar a este usuario", Toast.LENGTH_LONG).show()
            return
        }

        currentUser.idRol = newRoleId

        // Sincronizar correo y email para evitar errores 500 por nulos en el backend
        val tempCorreo = currentUser.correo
        val tempEmail = currentUser.email

        if (tempCorreo.isNullOrEmpty()) {
            if (!tempEmail.isNullOrEmpty()) {
                currentUser.correo = tempEmail
            }
        } else if (tempEmail.isNullOrEmpty()) {
            currentUser.email = tempCorreo
        }

        RetrofitClient.instance.actualizarUsuario(id, currentUser).enqueue(object : Callback<Usuario> {
            override fun onResponse(call: Call<Usuario>, response: Response<Usuario>) {
                if (response.isSuccessful) {
                    Toast.makeText(context, "Rol actualizado correctamente", Toast.LENGTH_SHORT).show()
                    requireActivity().onBackPressed()
                } else if (response.code() == 403) {
                    Toast.makeText(context, "Error de Jerarquía: Sin permisos", Toast.LENGTH_LONG).show()
                } else {
                    Toast.makeText(context, "Error: ${response.code()}", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<Usuario>, t: Throwable) {
                Toast.makeText(context, "Fallo de conexión", Toast.LENGTH_SHORT).show()
            }
        })
    }
}
